### PR TITLE
[serve] compact scheduling strategy

### PR DESCRIPTION
**Pull request by zcin**
[serve] compact scheduling strategy

Implement compact scheduling strategy when scheduling new pending replicas. This minimizes fragmentation.

This is feature flagged and can be turned on with `RAY_SERVE_USE_COMPACT_SCHEDULING_STRATEGY = 1`.

Signed-off-by: Cindy Zhang <cindyzyx9@gmail.com>

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/ray-project/ray/pull/43591).
* #43755
* __->__ #43591

### CODE REVIEW

**Comment by zcin** [comment]
@edoakes @jjyao Still writing tests but this is ready for an early first pass!

File: python/ray/serve/_private/deployment_scheduler.py [hunk 1]
```diff
@@ -338,11 +429,43 @@ def schedule(
...
+                scheduling_request
+                for replicas in self._pending_replicas.values()
+                for scheduling_request in replicas.values()
+            ]
+
+            replica_scheduling_requests.sort(key=key, reverse=True)
+
+            # Schedule each replica
+            for scheduling_request in replica_scheduling_requests:
+                self._schedule_replica_compactly(scheduling_request)
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_scheduler.py:462]: can we put this into a method like `_schedule_spread_deployment` to match `_schedule_compact_deployment`?

File: python/ray/serve/_private/deployment_scheduler.py [hunk 2]
```diff
@@ -338,11 +429,43 @@ def schedule(
...
+
+            # Flatten dict of deployment replicas into all replicas,
+            # then sort by decreasing resource size
+            replica_scheduling_requests = [
+                scheduling_request
+                for replicas in self._pending_replicas.values()
+                for scheduling_request in replicas.values()
+            ]
+
+            replica_scheduling_requests.sort(key=key, reverse=True)
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_scheduler.py:458]: suggest to pull this into a standalone method. quite hard to follow it inlined like this

File: python/ray/serve/_private/deployment_scheduler.py [hunk 3]
```diff
@@ -462,5 +585,95 @@ def key(node_and_num_running_replicas_of_all_deployments):
...
+            scheduling_request.placement_group_bundles
+            and scheduling_request.placement_group_strategy == "STRICT_PACK"
+        ):
+            bundle_resources = [
+                Resources(bundle)
+                for bundle in scheduling_request.placement_group_bundles
+            ]
+            required_resources = sum(bundle_resources, Resources())
+        else:
+            required_resources = Resources(scheduling_request.actor_resources)
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_scheduler.py:604]: this looks redundant with the sorting logic above -- possible to share the code for it?

File: python/ray/serve/_private/deployment_scheduler.py [hunk 4]
```diff
@@ -462,5 +585,95 @@ def key(node_and_num_running_replicas_of_all_deployments):
...
+
+        # Get node
+        target_node = self._find_best_available_node(
+            required_resources, self._get_available_resources_per_node()
+        )
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_scheduler.py:609]: no need to nest this call with `_schedule_replica_compactly`, I think instead the node can be chosen and passed in at the callsite.

btw, isn't the rest of this method call shared completely with the existing scheduling strategy? so it seems to me that we could abstract out just the logic to map replicas to nodes and the rest should not need to diverge across the strategies

**Review by edoakes — COMMENTED** [review]
Mostly stylistic comments because I was having some trouble following the code.

The algorithm looks reasonable for a first cut. I'll be interested to see how well this greedy approach works in practice.

File: python/ray/serve/_private/deployment_scheduler.py (hunk 3 above)
Comment (zcin) [reply python/ray/serve/_private/deployment_scheduler.py]: Moved to shared property of `ReplicaSchedulingRequest`.

File: python/ray/serve/_private/deployment_scheduler.py (hunk 4 above)
Comment (zcin) [reply python/ray/serve/_private/deployment_scheduler.py]: Good call! I've refactored the code so the scheduling code is shared in `_schedule_replica`, and the information that diverges is passed in at the callsite. PTAL!

File: python/ray/serve/_private/deployment_scheduler.py (hunk 2 above)
Comment (zcin) [reply python/ray/serve/_private/deployment_scheduler.py]: Makes sense, pulled out!

**Review by edoakes — COMMENTED** [review]
ok pretty much LGTM, let's get @jjyao's review too

File: .buildkite/serve.rayci.yml [hunk 5]
```diff
@@ -44,6 +44,20 @@ steps:
...
+      - serve
+      - python
+    instance_type: large
+    soft_fail: true
+    commands:
+      - bazel run //ci/ray_ci:test_in_docker -- //python/ray/serve/... //python/ray/tests/...  serve 
+        --except-tags post_wheel_build,gpu,worker-container,ha_integration
+        --workers "$${BUILDKITE_PARALLEL_JOB_COUNT}" --worker-id "$${BUILDKITE_PARALLEL_JOB}" --parallelism-per-worker 3
+        --build-name servebuild --test-env=EXPECTED_PYTHON_VERSION=3.9 --test-env=RAY_SERVE_USE_COMPACT_SCHEDULING_STRATEGY=1
+    depends_on: "servebuild"
```
[the remark this answers is not in the document]
Comment (zcin) [reply .buildkite/serve.rayci.yml:59]: OK, will change this to a smaller subset of tests.

File: python/ray/serve/_private/deployment_scheduler.py [hunk 6]
```diff
@@ -164,6 +181,12 @@ def is_non_strict_pack_pg(self) -> bool:
         )
 
 
+@dataclass
+class LaunchingReplicaInfo:
+    az: Optional[str] = None
+    node_id: Optional[str] = None
```
[the remark this answers is not in the document]
Comment (zcin) [reply python/ray/serve/_private/deployment_scheduler.py:187]: AZ is not used here, and `node_id` is only used for compact scheduling (vs normal spread scheduling). If both are None then it just means we're not using `launching_replicas` for any special scheduling. (launching replicas is also best effort, it contains the information that we "plan" on scheduling a new replica to before we get the round trip update from the GCS, but it's not guaranteed)

[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/deployment_scheduler.py:187]: az being optional makes sense

but for node_id, shouldn't it always be populated but just ignored by spread scheduling? or else this entire data structure unused?

[the remark this answers is not in the document]
Comment (zcin) [reply python/ray/serve/_private/deployment_scheduler.py:187]: Hmm, for spread scheduling the Serve controller doesn't look at the current cluster view / available nodes at all and instead lets core select a node that satisfies requirements (like resources, PG, etc), so we can't really populate it with anything (and also the calculations in `get_available_resources_per_node` will be inaccurate if we just randomly populate it). Whereas with compact scheduling we are selecting a node ourselves which is why we populate it.

I think the options here are set it to `None` or set it to `LaunchingReplicaInfo(az=None, node_id=None)`. Would you prefer the former?

[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/deployment_scheduler.py:187]: I see now, because this is populated in the shared method that both codepaths go through, but for spread case we won't actually target a node. In this I'm OK with it as-is, but I would suggest documenting that these are both best-effort and perhaps renaming the fields to something like `target_node_id` and `target_az`.

File: python/ray/serve/_private/deployment_scheduler.py [hunk 7]
```diff
@@ -294,6 +331,75 @@ def _get_node_to_running_replicas(
 
         return res
 
+    def _get_available_resources_per_node(self) -> Dict[str, Resources]:
```
[the remark this answers is not in the document]
Comment (zcin) [reply python/ray/serve/_private/deployment_scheduler.py:358]: added!

File: python/ray/serve/_private/deployment_scheduler.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zcin) [reply python/ray/serve/_private/deployment_scheduler.py]: Makes sense, I've added missing docstrings and renamed them to `target_node_id` and `target_az`!

**Comment by zcin** [comment]
@jjyao Ping for review 🙏

File: python/ray/serve/_private/deployment_scheduler.py [hunk 8]
```diff
@@ -127,6 +132,26 @@ class ReplicaSchedulingRequest:
     placement_group_strategy: Optional[str] = None
     max_replicas_per_node: Optional[int] = None
 
+    @property
+    def required_resources(self) -> Resources:
+        """The resources required to schedule this replica on a node.
+
+        If this replicas uses a strict pack placement group, the
```
Comment (jjyao) [inline python/ray/serve/_private/deployment_scheduler.py:139]: this replica

File: python/ray/serve/_private/deployment_scheduler.py [hunk 9]
```diff
@@ -174,16 +222,21 @@ def __init__(
...
         self._deployments: Dict[DeploymentID, DeploymentSchedulingInfo] = {}
         # Replicas that are waiting to be scheduled.
         # {deployment_id: {replica_name: deployment_upscale_request}}
-        self._pending_replicas = defaultdict(dict)
+        self._pending_replicas: Dict[
+            DeploymentID, Dict[str, ReplicaSchedulingRequest]
+        ] = defaultdict(dict)
         # Replicas that are being scheduled.
         # The underlying actors have been submitted.
         # {deployment_id: {replica_name: target_node_id}}
```
Comment (jjyao) [inline python/ray/serve/_private/deployment_scheduler.py:236]: Update the comment

File: python/ray/serve/_private/deployment_scheduler.py [hunk 10]
```diff
@@ -174,16 +222,21 @@ def __init__(
         self,
         cluster_node_info_cache: ClusterNodeInfoCache,
         head_node_id: str,
+        create_placement_group: Callable,
```
Comment (jjyao) [inline python/ray/serve/_private/deployment_scheduler.py:225]: Nit: create_placement_group_fn ?

File: python/ray/serve/_private/deployment_scheduler.py [hunk 11]
```diff
@@ -280,6 +333,17 @@ def on_replica_recovering(
 
         self._recovering_replicas[deployment_id].add(replica_name)
 
+    def _on_replica_launching(
+        self,
+        deployment_id: DeploymentID,
+        replica_name: str,
+        target_node_id: str = None,
```
Comment (jjyao) [inline python/ray/serve/_private/deployment_scheduler.py:340]: Optional[str]

File: python/ray/serve/_private/deployment_scheduler.py [hunk 12]
```diff
@@ -294,6 +358,95 @@ def _get_node_to_running_replicas(
...
+        available_resources = (
+            self._cluster_node_info_cache.get_available_resources_per_node()
+        )
+        total_resources = self._cluster_node_info_cache.get_total_resources_per_node()
+
+        gcs_info = {node_id: Resources(r) for node_id, r in available_resources.items()}
+
+        # Manually calculate available resources per node by subtracting
+        # launching and running replicas from total resources
+        available_minus_replicas = {
```
Comment (jjyao) [inline python/ray/serve/_private/deployment_scheduler.py:391]: total_minus_replicas?

File: python/ray/serve/_private/deployment_scheduler.py [hunk 13]
```diff
@@ -294,6 +358,95 @@ def _get_node_to_running_replicas(
...
+        def custom_min(a: Resources, b: Resources):
+            keys = set(a.keys()) | set(b.keys())
+            res = Resources()
+            for key in keys:
+                res[key] = min(a.get(key), b.get(key))
+            return res
+
+        # Filter by active node ids (alive but not draining)
+        return {
+            node_id: custom_min(gcs_info[node_id], available_minus_replicas[node_id])
```
Comment (jjyao) [inline python/ray/serve/_private/deployment_scheduler.py:416]: Will we have KeyError?

**Review by jjyao — COMMENTED** [review]
Will continue.

File: python/ray/serve/_private/deployment_scheduler.py (hunk 13 above)
Comment (zcin) [reply python/ray/serve/_private/deployment_scheduler.py]: I think not for `total_minus_replicas` because `get_active_node_ids` comes from the same cached dictionary as `total_minus_replicas`. However `gcs_info` makes a separate call to GCS, so I guess there is a slight chance of key error here. Defensively changed it to use an empty resource if `node_id` doesn't exist.

File: python/ray/serve/_private/deployment_scheduler.py [hunk 14]
```diff
@@ -294,6 +358,98 @@ def _get_node_to_running_replicas(
...
+        This strategy picks the node where, if the required resources
+        were to be scheduled on that node, it will leave the smallest
+        remaining space. This minimizes fragmentation of resources.
+        """
+
+        min_remaining_space = Resources(
+            {
+                "CPU": float("inf"),
+                "GPU": float("inf"),
+                "memory": float("inf"),
```
Comment (jjyao) [inline python/ray/serve/_private/deployment_scheduler.py:437]: We should consider all resources of `required_resources`?

File: python/ray/serve/_private/deployment_scheduler.py [hunk 15]
```diff
@@ -311,8 +467,75 @@ def schedule(
...
         raise NotImplementedError
 
+    def _schedule_replica(
+        self,
+        scheduling_request: ReplicaSchedulingRequest,
+        default_scheduling_strategy: str,
+        target_labels: Dict[str, Any] = None,
+        target_node_id: str = None,
+    ):
+        """Schedule a replica from a scheduling request."""
```
Comment (jjyao) [inline python/ray/serve/_private/deployment_scheduler.py:477]: Could you add docstring for the parameters.

File: python/ray/serve/_private/deployment_scheduler.py [hunk 16]
```diff
@@ -311,8 +467,75 @@ def schedule(
         """
         raise NotImplementedError
 
+    def _schedule_replica(
+        self,
+        scheduling_request: ReplicaSchedulingRequest,
+        default_scheduling_strategy: str,
+        target_labels: Dict[str, Any] = None,
```
Comment (jjyao) [inline python/ray/serve/_private/deployment_scheduler.py:474]: Optional

File: python/ray/serve/_private/deployment_scheduler.py [hunk 17]
```diff
@@ -311,8 +467,75 @@ def schedule(
...
+        )
+
+        if isinstance(scheduling_strategy, PlacementGroupSchedulingStrategy):
+            placement_group = scheduling_strategy.placement_group
+
+        scheduling_request.on_scheduled(actor_handle, placement_group=placement_group)
+
     @abstractmethod
-    def detect_compact_opportunities(self) -> Tuple[str, float]:
+    def detect_compact_opportunities(self) -> Tuple[Optional[str], Optional[float]]:
```
Comment (jjyao) [inline python/ray/serve/_private/deployment_scheduler.py:538]: should it be `Optional[Tuple[str, float]]`

File: python/ray/serve/_private/deployment_scheduler.py [hunk 18]
```diff
@@ -311,8 +467,75 @@ def schedule(
...
+
+        actor_options = copy.copy(scheduling_request.actor_options)
+        if scheduling_request.max_replicas_per_node is not None:
+            if "resources" not in actor_options:
+                actor_options["resources"] = {}
+            # Using implicit resource (resources that every node
+            # implicitly has and total is 1)
+            # to limit the number of replicas on a single node.
+            actor_options["resources"][
+                f"{ray._raylet.IMPLICIT_RESOURCE_PREFIX}"
```
Comment (jjyao) [inline python/ray/serve/_private/deployment_scheduler.py:518]: node resource doesn't handle implicit resource?

**Review by jjyao — COMMENTED** [review]
LG!

File: python/ray/serve/_private/deployment_scheduler.py (hunk 15 above)
Comment (zcin) [reply python/ray/serve/_private/deployment_scheduler.py:477]: done!

**Review by jjyao — APPROVED** [review]

File: python/ray/serve/_private/deployment_scheduler.py (hunk 14 above)
Comment (zcin) [reply python/ray/serve/_private/deployment_scheduler.py]: Fixed!

File: python/ray/serve/_private/deployment_scheduler.py (hunk 17 above)
Comment (zcin) [reply python/ray/serve/_private/deployment_scheduler.py:538]: Will address in https://github.com/ray-project/ray/pull/43755
